### PR TITLE
Adopt a new way to go with the loglevels

### PR DESCRIPTION
**Pull request by DevUt**

### CODE REVIEW

File: C/kmsg_log.c [hunk 1]
```diff
@@ -13,14 +13,14 @@
...
- char *buf = malloc(sizeof(char) * 256);
+ char *buf = malloc(256);
  int ret = -1;
 
   // Write to kmsg if possibe.
  int outbuf = open("/dev/kmsg", O_WRONLY);
   if (outbuf > 0) // Trying to open KMSG,
   {
-    strcpy(buf,*msg);
+    strcpy(buf,msg);
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: btw we still didn't check msg != NULL

File: C/kmsg_log.c [hunk 2]
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
...
-		return(ret);
-	}
+/*
+  This functions prefixes the Warning label
+  to the real message and sends it to klog
+  for real "printing" to kmsg (IF LOG_LOC not defined or used)
+*/
+int logwarn(char *warning)
+{
+ char *warn = malloc(256);
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: Still bad indent :s

File: C/kmsg_log.c [hunk 3]
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
...
+  strcpy(warn ,"Warning!: ");
+  strcat(warn ,warning);
+ int ret = klog(warn) ;
+  return(ret);
+  free(warn);
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: Unreachable code

File: C/kmsg_log.c [hunk 4]
```diff
@@ -13,14 +13,14 @@
 int klog(char *msg)
 {
 
- char *buf = malloc(sizeof(char) * 256);
- int ret = -1; 
+ char *buf = malloc(256);
+ int ret = -1;
 
   // Write to kmsg if possibe.
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: You didn't remove comments anymore?

File: C/kmsg_log.c [hunk 5]
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
...
+  free(error);
+ }
+  /*
+    This functions prefixes the debug label
+    to the real message and sends it to klog
+    for real "printing" to kmsg (IF LOG_LOC not defined or used)
+  */
+int logdebug (char *debug_msg)
+{
+ char *debug = malloc(256);
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: Your are using malloc everywhere, good catch. 
But you set the same value on all entries. Can you use a preprocessor variable instead please?

**Review by nailyk-fr — CHANGES_REQUESTED** [review]
Really hard to review. Patch is unreadable because of the indent problem. 
Your new version looks good except the indent.

**Review by Rosalie241 — APPROVED** [review]

File: C/kmsg_log.c [hunk 6]
```diff
@@ -8,102 +8,64 @@
...
+  if (outbuf > 0)
   {
-    strcpy(buf,msg);
-  } else { //If kmesg don't open write to sd
-    /*
-      Now if the file doesn't exists, we would open another file on the sdcard to transmit our KMSG.
-    */
-    if (access(LOG_LOC, F_OK ) == -1) {  // Check if the file exists.
-      outbuf = open(LOG_LOC, O_CREAT | O_WRONLY); // If it doesn't exist, create it and open it to write to it.
+    switch (level)
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: Switch misplaced, need to be out of the if(open())else statement.

File: C/kmsg_log.c [hunk 7]
```diff
@@ -8,102 +8,64 @@
...
+      case 1 : tag = "Warn :";
+                strcat(tag,msg);
+              break ;
+      case 2 : tag = "Error :"
+                strcat(tag,msg);
+              break ;
+      case 3 : tag = "Debug :"
+                strcat(tag,msg);
+              break ;
+      default : tag = "Error : Unknown log level "
```
Comment (nailyk-fr) [inline C/kmsg_log.c]: Please two point: 
Make every log level same string lenght. (like ERR, WAR, INF)
I think you can import kernel log level too and assume default as max: 

case 7: 
default: 
    foo bar; 
    break;

File: C/kmsg_log.c [hunk 8]
```diff
@@ -8,102 +8,64 @@
...
     }
+  }
+	ret = write(outbuf, tag, strlen(tag) );
+	close(outbuf);
 
-	ret = write(outbuf, buf, strlen(buf) ); // Write "Test" to KMSG
-
-	close(outbuf); // CLOSE THE FILE DISCRIPTOR TO AVOID LEAK ;
-
-	free(buf); // Empty buf memory.
```
Comment (nailyk-fr) [inline C/kmsg_log.c:50]: no free(tag)?

**Review by nailyk-fr — COMMENTED** [review]
This one looks good, except the misplaced switch statement. 
Please, only for my happiness, can you rename tag --> buf?

File: C/kmsg_log.c (hunk 7 above)
Comment (DevUt) [reply C/kmsg_log.c]: I didn't get the default point please elaborate

File: C/kmsg_log.c (hunk 8 above)
Comment (DevUt) [reply C/kmsg_log.c:50]: Why free ? 
I used stack memory instead of heap 
I have no control on it (and its good) 
Its premanaged

Comment (nailyk-fr) [reply C/kmsg_log.c:50]: Forget it. Yes and this way is way better!

File: C/kmsg_log.c (hunk 7 above)
Comment (nailyk-fr) [reply C/kmsg_log.c]: Don't mind, will push them afterward.

File: C/kmsg_log.c [hunk 9]
```diff
@@ -19,28 +24,28 @@
...
+  and writes "COULD NOT OPEN KMSG PLEASE FIX".
 */
 int klog(int level, char *msg)
 {
 
  int ret = -1;
  char buf [256];
 
- int outbuf = open("/dev/kmsg", O_WRONLY);
+ int outbuf = open(DEFAULT_LOG_LOC, O_WRONLY);
```
Comment (nailyk-fr) [inline C/kmsg_log.c:38]: Force define of DEFAULT_LOG_LOC in case not passed to makefile.

File: C/Android.mk [hunk 10]
```diff
@@ -2,4 +2,8 @@ LOCAL_PATH := $(call my-dir)
 ifeq ($(LOG_WRAP_TAGS),true)
 	LOCAL_CFLAGS += -DLOG_WRAP_TAGS
 endif
+
+ifeq ($(DEFAULT_LOG_LOC),)
+	LOCAL_CFLAGS += -DDEFAULT_LOG_LOC=\"$(DEFAULT_LOG_LOC)\"
```
Comment (nailyk-fr) [inline C/Android.mk:7]: Add possibility to define ALT_LOG_LOC too.

**Review by nailyk-fr — APPROVED** [review]

**Pull request merged**